### PR TITLE
client_pubkey always leaks in server_process_dh()

### PR DESCRIPTION
**Pull request by Sashan**
This fixes the double free of `dh_server` I've introduced in earlier pull request of this branch.

### CODE REVIEW

File: src/plugins/preauth/pkinit/pkinit_crypto_openssl.c [hunk 1]
```diff
@@ -2970,24 +2975,27 @@ server_process_dh(krb5_context context,
     pub_key = BN_to_ASN1_INTEGER(server_pubkey, NULL);
     if (pub_key == NULL)
         goto cleanup;
-    *dh_pubkey_len = i2d_ASN1_INTEGER(pub_key, NULL);
-    if ((p = *dh_pubkey = malloc(*dh_pubkey_len)) == NULL)
+    dh_pubkey_len = i2d_ASN1_INTEGER(pub_key, NULL);
+    if ((p = dh_pubkey = malloc(dh_pubkey_len)) == NULL)
```
Comment (frozencemetery) [inline src/plugins/preauth/pkinit/pkinit_crypto_openssl.c]: While you're touching this line, please move the function call outside of the `if`'s conditional.

File: src/plugins/preauth/pkinit/pkinit_crypto_openssl.c [hunk 2]
```diff
@@ -2970,24 +2975,27 @@ server_process_dh(krb5_context context,
...
+    server_key = NULL;
+    *server_key_len_out = server_key_len;
 
+    retval = 0;
 cleanup:
-    BN_free(client_pubkey);
     DH_free(dh_server);
-    free(*dh_pubkey);
-    free(*server_key);
+    BN_free(client_pubkey);
```
Comment (frozencemetery) [inline src/plugins/preauth/pkinit/pkinit_crypto_openssl.c]: Is there a reason this line moves?

File: src/plugins/preauth/pkinit/pkinit_crypto_openssl.c [hunk 3]
```diff
@@ -2911,20 +2911,25 @@ server_process_dh(krb5_context context,
...
     const BIGNUM *server_pubkey;
+    unsigned char *dh_pubkey, *server_key;
+    unsigned int dh_pubkey_len, server_key_len;
 
-    *dh_pubkey = *server_key = NULL;
-    *dh_pubkey_len = *server_key_len = 0;
+    *dh_pubkey_out = *server_key_out = NULL;
+    *dh_pubkey_len_out = *server_key_len_out = 0;
+
+    dh_pubkey = server_key = NULL;
```
Comment (frozencemetery) [inline src/plugins/preauth/pkinit/pkinit_crypto_openssl.c]: Please do these initializations in the same way the surrounding code does for local variables.

**Comment by greghudson** [comment]
Closing this as it was addressed in PR #736.

**Comment by greghudson** [comment]
Wait, sorry, reopening; I forgot this was separate from the backport commit.

**Comment by greghudson** [comment]
I pushed a revised commit, mainly to also change client_create_dh() and client_process_dh().  Robbie, please have a look if you have time.

**Pull request merged**